### PR TITLE
NFT nonce per lastByte

### PR DESCRIPTION
**Pull request by sasurobert**
NFT creation on multiple shards

### CODE REVIEW

File: builtInFunctions/esdtRoles.go [hunk 1]
```diff
@@ -76,6 +87,15 @@ func (e *esdtRoles) ProcessBuiltinFunction(
 	return vmOutput, nil
 }
 
+// Nonces on multi shard NFT create ar from (LastByte * MaxUint64 / 256), this is in order to differentiate them
```
Comment (bogdan-rosianu) [inline builtInFunctions/esdtRoles.go:90]: are*

File: builtInFunctions/esdtRoles.go [hunk 2]
```diff
@@ -134,14 +154,21 @@ func (e *esdtRoles) CheckAllowedToExecute(account vmcommon.UserAccountHandler, t
...
 		return ErrActionNotAllowed
 	}
+	if !searchForRole(roles, action) {
+		return ErrActionNotAllowed
+	}
+
+	return nil
+}
 
+func searchForRole(roles *esdt.ESDTRoles, action []byte) bool {
```
Comment (bogdan-rosianu) [inline builtInFunctions/esdtRoles.go:164]: could rename this function to `isRoleSet()` or `isActionAllowed()` as the function returns a boolean

File: builtInFunctions/esdtRoles.go [hunk 3]
```diff
@@ -67,6 +68,16 @@ func (e *esdtRoles) ProcessBuiltinFunction(
 		deleteRoles(roles, vmInput.Arguments[1:])
 	}
 
+	for _, arg := range vmInput.Arguments[1:] {
+		if bytes.Equal(arg, []byte(core.ESDTRoleNFTCreateMultiShard)) {
+			err = saveLatestNonce(acntDst, vmInput.Arguments[0], computeStartNonce(vmInput.RecipientAddr))
+			if err != nil {
+				return nil, err
+			}
+			break
```
Comment (bogdan-rosianu) [inline builtInFunctions/esdtRoles.go:77]: empty line before `break` ?

File: builtInFunctions/esdtRoles.go [hunk 4]
```diff
@@ -67,6 +68,16 @@ func (e *esdtRoles) ProcessBuiltinFunction(
 		deleteRoles(roles, vmInput.Arguments[1:])
 	}
 
+	for _, arg := range vmInput.Arguments[1:] {
+		if bytes.Equal(arg, []byte(core.ESDTRoleNFTCreateMultiShard)) {
```
Comment (bogdan-rosianu) [inline builtInFunctions/esdtRoles.go:72]: could refactor:
```
if !bytes.Equal(..) {
    continue
}

    err = saveLatestNonce(....
...
```

**Review by bogdan-rosianu — CHANGES_REQUESTED** [review]

File: builtInFunctions/esdtRoles.go [hunk 5]
```diff
@@ -76,6 +87,15 @@ func (e *esdtRoles) ProcessBuiltinFunction(
 	return vmOutput, nil
 }
 
+// Nonces on multi shard NFT create ar from (LastByte * MaxUint64 / 256), this is in order to differentiate them
+// even like this, if one contract makes 1000 NFT create on each block, it would need 14 million years to occupy the whole space
+// 2 ^ 64 / 256 / 1000 / 14400 / 365 ~= 14 million
+func computeStartNonce(destAddress []byte) uint64 {
```
Comment (iulianpascalau) [inline builtInFunctions/esdtRoles.go:93]: this function will return only 256 possible values. Is this ok? I think we should have added a current nonce to the computed startNonce

**Review by iulianpascalau — CHANGES_REQUESTED** [review]

File: builtInFunctions/esdtRoles.go (hunk 5 above)
Comment (sasurobert) [reply builtInFunctions/esdtRoles.go:93]: This is OK. this is the startNonce to be computed per lastAddress when multishardNFTCreate role is set.

File: builtInFunctions/esdtRoles.go (hunk 1 above)
Comment (sasurobert) [reply builtInFunctions/esdtRoles.go:90]: done

File: builtInFunctions/esdtRoles.go (hunk 2 above)
Comment (sasurobert) [reply builtInFunctions/esdtRoles.go:164]: deleted as we had another function which did the same

File: builtInFunctions/esdtRoles.go (hunk 3 above)
Comment (sasurobert) [reply builtInFunctions/esdtRoles.go:77]: done

File: builtInFunctions/esdtRoles.go (hunk 4 above)
Comment (sasurobert) [reply builtInFunctions/esdtRoles.go:72]: done

**Review by bogdan-rosianu — DISMISSED** [review]

**Review by iulianpascalau — APPROVED** [review]

**Review by bogdan-rosianu — APPROVED** [review]

**Review by bogdan-rosianu — APPROVED** [review]

**Review by iulianpascalau — APPROVED** [review]

**Pull request merged**